libs/algorithms/tests/regressions/count_3646.cpp

```diff
@@ -39,6 +39,32 @@ struct BitCountingIterator : public Iterator<std::int64_t, stopValue>
         return countBits(this->state + n);
     }
 
+    BitCountingIterator& operator++()
+    {
+        ++(this->state);
+        return *this;
+    }
+
+    BitCountingIterator operator++(int)
+    {
+        auto copy = *this;
+        ++(*this);
+        return copy;
+    }
+
+    BitCountingIterator& operator--()
+    {
+        --(this->state);
+        return *this;
+    }
+
+    BitCountingIterator operator--(int)
+    {
+        auto copy = *this;
+        --(*this);
+        return copy;
+    }
+
```

Comment:
This test curiously started failing. It was just lucky with the old chunk size though. For some reason the parallel policy lead to the superclass (`Iterator`) being dereferenced ([here](https://github.com/STEllAR-GROUP/hpx/blob/3df4e4489ff5a360a5838f88b3782489c614827f/libs/algorithms/include/hpx/parallel/util/loop.hpp#L172-L175), the first one dereferences correctly and the others don't). Can someone see why it wouldn't use the dereference operator from `BitCountingIterator`?

Reply:
I'll have a look

Reply:
Thank you! Note that I've force pushed in the meantime.

Reply:
@msimberg I'm not able to reproduce the issue locally :/ No idea what's wrong...

Reply:
I should've said it explicitly, but these changes fix the problem. I just don't understand why it doesn't work correctly without. Did you run the latest commit on this branch?

Reply:
The number of threads also affects the result... because of its influence on the chunk size.

Reply:
Yes, I ran the latest commit on this branch.

Reply:
Ok, you added `operator++()` and friends. Without them the code was using the corresponding operators implemented by the base class. Those however return `Iterator` which caused all subsequent operations to directly work with the base class only...

Reply:
Uhh, right. For some reason I expected them to return a `BitCountingIterator` cast to `Iterator`, but looking at it again I'm not sure how that would ever happen. Then this makes more sense, thanks!